### PR TITLE
Batch load cache

### PR DESCRIPTION
**Pull request by xchen1189**
## Checklist

Please make sure the following requirements are complete:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes /
  features)
- [ ] All automated checks pass (CI/CD)

### CODE REVIEW

File: lib/fast_jsonapi/serialization_core.rb [hunk 1]
```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
...
+        end
+
         options[:namespace] ||= 'jsonapi-serializer'
 
-        fieldset_key = fieldset.join('_')
+        return options unless fieldset
 
         # Use a fixed-length fieldset key if the current length is more than
         # the length of a SHA1 digest
+        fieldset_key = fieldset.join('-')
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: why?

File: lib/fast_jsonapi/serialization_core.rb [hunk 2]
```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
...
         if fieldset_key.length > 40
           fieldset_key = Digest::SHA1.hexdigest(fieldset_key)
         end
 
-        options[:namespace] = "#{options[:namespace]}-fieldset:#{fieldset_key}"
+        options[:namespace] = [options[:namespace], fieldset_key].join(':')
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: is this change necessary?

File: lib/fast_jsonapi/serialization_core.rb (hunk 1 above)
Comment (xchen1189) [reply lib/fast_jsonapi/serialization_core.rb]: Why use dash over underscore? It's a matter of taste, also want to consistent with cache_key style like `meal_instances/321724-20210617211745378854`.

File: lib/fast_jsonapi/serialization_core.rb [hunk 3]
```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
...
-          cache_opts = record_cache_options(cache_store_options, fieldset, includes_list, params)
-          record_hash = cache_store_instance.fetch(record, **cache_opts) do
-            temp_hash = id_hash(id_from_record(record, params), record_type, true)
-            temp_hash[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
-            temp_hash[:relationships] = relationships_hash(record, cachable_relationships_to_serialize, fieldset, includes_list, params) if cachable_relationships_to_serialize.present?
-            temp_hash[:links] = links_hash(record, params) if data_links.present?
-            temp_hash
+          cache_opts = record_cache_options(record, cache_store_options, fieldset, includes_list, params)
+          name_space = cache_opts.delete(:namespace)
+          cache_key = generate_cache_key(record, name_space)
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb:79]: how is this different from the default ASupport implementation?

File: lib/fast_jsonapi/serialization_core.rb (hunk 2 above)
Comment (xchen1189) [reply lib/fast_jsonapi/serialization_core.rb]: no, not necessary, I'm really adopting to chowbus style of building cache keys.

File: lib/fast_jsonapi/serialization_core.rb [hunk 4]
```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
...
+          klass = self
+          param = {
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]:
```suggestion
          fetch_query = {
```
`param` and `params` are a bit confusing

File: lib/fast_jsonapi/serialization_core.rb (hunk 2 above)
Comment (xchen1189) [reply lib/fast_jsonapi/serialization_core.rb]: since we want to minimum the change set. I can undo those style changes.

File: lib/fast_jsonapi/serialization_core.rb (hunk 3 above)
Comment (zinosama) [reply lib/fast_jsonapi/serialization_core.rb:79]: i see. AS doesn't support multi read with different namespaces.

Comment (zinosama) [reply lib/fast_jsonapi/serialization_core.rb:79]: maybe add a comment here explain why we don't use the built-in namespace option

Comment (xchen1189) [reply lib/fast_jsonapi/serialization_core.rb:79]: yep

File: lib/fast_jsonapi/serialization_core.rb [hunk 5]
```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
...
+              .reject { |h| cache_hits.keys.include? h[:cache_key] }
+              .group_by { |h| h[:cache_opts] }
+            # replace grouped batch param with Hash: { cache_key_1 => record_hash_1, ... }
+            uncached_by_opts.transform_values! do |arr|
+              arr.each_with_object({}) do |b_param, record_hash_pairs|
+                temp_hash = b_param[:klass].id_hash(b_param[:klass].id_from_record(b_param[:record], b_param[:params]), b_param[:klass].record_type, true)
+                temp_hash[:attributes] = b_param[:klass].attributes_hash(b_param[:record], fieldset, b_param[:params]) if b_param[:klass].attributes_to_serialize.present?
+                temp_hash[:relationships] = b_param[:klass].relationships_hash(b_param[:record], nil, fieldset, includes_list, b_param[:params]) if b_param[:klass].relationships_to_serialize.present?
+                temp_hash[:links] = b_param[:klass].links_hash(b_param[:record], b_param[:params]) if b_param[:klass].data_links.present?
+                temp_hash[:meta] = b_param[:klass].meta_hash(b_param[:record], b_param[:params]) if b_param[:klass].meta_to_serialize.present?
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: add a TODO comment here

File: lib/fast_jsonapi/serialization_core.rb (hunk 1 above)
Comment (zinosama) [reply lib/fast_jsonapi/serialization_core.rb]: not worth a breaking change for the gem

File: lib/fast_jsonapi/serialization_core.rb [hunk 6]
```diff
@@ -65,46 +66,90 @@ def meta_hash(record, params = {})
...
+          ::BatchLoader.for(param).batch do |batch_params, loader|
+            cache_keys = batch_params.map { |h| h[:cache_key] }
+            # load the cached value from cache store
+            cache_hits = cache_store_instance.read_multi(cache_keys)
+
+            # for not cached record, group it by cache options
+            uncached_by_opts = batch_params
+              .reject { |h| cache_hits.keys.include? h[:cache_key] }
+              .group_by { |h| h[:cache_opts] }
+            # replace grouped batch param with Hash: { cache_key_1 => record_hash_1, ... }
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: this comment will be more useful if we show the overall structure of uncached_by_opts instead of only the value:
```
{
  {...cache_option} => {
    "cache_key_1" => {...record_hash_1},
    "cache_key_2"=> {...record_hash_2},
    ...
  },
  .... # different cache options
}
```

File: lib/fast_jsonapi/serialization_core.rb [hunk 7]
```diff
@@ -65,46 +66,90 @@ def meta_hash(record, params = {})
...
+                temp_hash = b_param[:klass].id_hash(b_param[:klass].id_from_record(b_param[:record], b_param[:params]), b_param[:klass].record_type, true)
+                temp_hash[:attributes] = b_param[:klass].attributes_hash(b_param[:record], fieldset, b_param[:params]) if b_param[:klass].attributes_to_serialize.present?
+                temp_hash[:relationships] = b_param[:klass].relationships_hash(b_param[:record], nil, fieldset, includes_list, b_param[:params]) if b_param[:klass].relationships_to_serialize.present?
+                temp_hash[:links] = b_param[:klass].links_hash(b_param[:record], b_param[:params]) if b_param[:klass].data_links.present?
+                temp_hash[:meta] = b_param[:klass].meta_hash(b_param[:record], b_param[:params]) if b_param[:klass].meta_to_serialize.present?
+                record_hash_pairs[b_param[:cache_key]] = temp_hash
+              end
+            end
+            # cache the uncached record
+            uncached_by_opts.each do |opts, record_hash_pairs|
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: `record_hash_pairs` is misleading - this is not a pair of record hashes, but rather `record_hashes_by_cache_key `

Comment (zinosama) [reply lib/fast_jsonapi/serialization_core.rb]:
```suggestion
            uncached_by_opts.each do |cache_opts, record_hashes_by_cache_key |
```

**Comment by zinosama** [comment]
this looks good. just some rename suggestions

File: lib/fast_jsonapi/serialization_core.rb (hunk 3 above)
Comment (nebolsin) [reply lib/fast_jsonapi/serialization_core.rb:79]: Can you please elaborate a bit on why do we need to use different namespaces?

Comment (nebolsin) [reply lib/fast_jsonapi/serialization_core.rb:79]: Okay, I see. For some mysterious reason `jsonapi-serializers` uses CacheStore's parameter `namespace` to distinguish cache keys for a same endpoint and data queried with different fieldsets, instead of making it a part of the key. That's quite surprising and certainly unexpected, but still I don't think it's possible to get several different values for namespaces in one particular serialiser processing a single requests.

Comment (xchen1189) [reply lib/fast_jsonapi/serialization_core.rb:79]: good call, that case it's also quite interesting. The main reason is the params options passed in serializer could result different values. For example, you can check endpoints `Api::V2::MealInstanceController#fetch` and `Api::OnDemand::V2::MealInstanceController#fetch`.

File: lib/fast_jsonapi/object_serializer.rb [hunk 8]
```diff
@@ -34,11 +34,13 @@ def initialize(resource, options = {})
...
-      if self.class.is_collection?(@resource, @is_collection)
-        return hash_for_collection
+      result = if self.class.is_collection?(@resource, @is_collection)
+        hash_for_collection
+      else
+        hash_for_one_record
       end
-
-      hash_for_one_record
+      return deep_symbolize_keys(result) if self.class.cache_store_instance
```
Comment (xchen1189) [inline lib/fast_jsonapi/object_serializer.rb]: manually break the batchloader promise to gain some performance benefits. 

cannot simple call as_json because the chowbus response-serializer won't accept hash with string key, and `#with_indifferent_access` shallow symbolized the hash keys.

**Review by zinosama — APPROVED** [review]

File: lib/fast_jsonapi/object_serializer.rb [hunk 9]
```diff
@@ -34,11 +34,13 @@ def initialize(resource, options = {})
...
-      if self.class.is_collection?(@resource, @is_collection)
-        return hash_for_collection
+      result = if self.class.is_collection?(@resource, @is_collection)
+        hash_for_collection
+      else
+        hash_for_one_record
       end
-
-      hash_for_one_record
+      return self.class.deep_sync(result) if self.class.cache_store_instance
```
Comment (zinosama) [inline lib/fast_jsonapi/object_serializer.rb]: is this gonna cause problem for action text serializer in production?

**Review by zinosama — APPROVED** [review]

File: lib/fast_jsonapi/serialization_core.rb [hunk 10]
```diff
@@ -101,18 +101,27 @@ def record_hash(record, fieldset, includes_list, params = {})
...
-                record_hash[:links] = b_param[:klass].links_hash(b_param[:record], b_param[:params]) if b_param[:klass].data_links.present?
-                record_hashes_by_cache_key[b_param[:cache_key]] = record_hash
+            uncached_by_opts = nil
+            Datadog.tracer.trace('uncached_by_opts', resource: 'CacheSerilization') do
+              uncached_by_opts = batch_params
+                .reject { |h| cache_hits[h[:cache_key]] }
+                .group_by { |h| h[:cache_opts] }
+            end
+
+            Rails.logger.info('serializer_cache_stats', cache_hits: cache_hits.size, cache_miss: cache_keys - cache_hits.size)
```
Comment (zinosama) [inline lib/fast_jsonapi/serialization_core.rb]: `cache_keys.size - cache_hits.size`

**Review by zinosama — APPROVED** [review]
